File: src/api/lifecycle.ts
// Copyright (c) 2026 OpenBKN. All rights reserved.
// Licensed under the Apache License, Version 2.0. See the LICENSE file in the project root.

/**
 * Managed-lifecycle sessions for the context-loader's HTTP `/kn/*` surface.
 *
 * Deploys from 0.1.3 on reject every POST under `/kn/` whose body omits
 * `bkn_context`, and the MCP tool surface wants the same object in its tool
 * arguments. A caller with no business conversation of its own — a CLI
 * invocation, a script — therefore fails closed on both. Core will not hand a
 * conversation to an end user directly either: its REST surface demands a
 * trusted gateway identity. The context-loader's MCP tools are that gateway, so
 * the route to a usable `bkn_context` runs through them, and this module walks
 * it.
 *
 * An earlier draft of this file claimed MCP was exempt because the server
 * merges one conversation per connection. That fallback exists, but it is a
 * fallback: a connection that supplies its own context is taken at its word,
 * and supplying one is what makes the evidence land on a nameable turn rather
 * than an anonymous per-connection bucket.
 *
 * The contract (context-loader `mcp.yaml`) is two tools: `bkn_start_interaction`
 * mints the conversation and interaction ids, `bkn_finish_interaction` ends the
 * interaction. `bkn_context` is `BKNContext` (`additionalProperties: false`): the
 * two ids plus the optional causality fields. Operation identity is derived
 * server-side, so callers never send an `operation_key`.
 *
 * A deploy whose catalog lacks `bkn_start_interaction` needs no context; sending
 * none keeps working exactly as before. Once the tool is advertised, however, a
 * failed handshake is authoritative and is surfaced.
 */
import { createHash } from "node:crypto";
import type { RequestContext } from "../types.js";
import { HttpError, InputError, ToolError } from "../utils/errors.js";
import { callToolRaw, mcpInfo } from "./context-loader.js";
import { inheritVersionCheck } from "./version-check.js";

/**
 * The body field the lifecycle middleware reads. Snake_case: it goes on the wire.
 *
 * Mirrors `BKNContext` in the context-loader contract (`_shared/common.yaml`),
 * which is `additionalProperties: false`.
 */
export interface BknContext {
  conversation_id: string;
  interaction_id: string;
  parent_operation_id?: string;
  causation_event_ids?: string[];
  business_refs?: Array<{ ref_type: string; ref_id: string; version?: string }>;
}

/** The keys `BKNContext` accepts. Anything else is rejected by a strict deploy. */
const WIRE_BKN_CONTEXT_KEYS = [
  "conversation_id",
  "interaction_id",
  "parent_operation_id",
  "causation_event_ids",
  "business_refs",
] as const;

/**
 * Reduce a caller-built `bkn_context` to the fields the contract accepts.
 *
 * Context Loader derives the Operation identity itself and `BKNContext` is
 * `additionalProperties: false`, so forwarding any other key — an
 * `operation_key` from an older caller included — would get the call refused.
 */
export function toWireBknContext(bknContext: object): BknContext {
  const source = bknContext as Record<string, unknown>;
  const wire: Record<string, unknown> = {};
  for (const key of WIRE_BKN_CONTEXT_KEYS) {
    if (source[key] !== undefined) wire[key] = source[key];
  }
  return wire as unknown as BknContext;
}

/** Whether a deploy speaks the managed lifecycle, decided from its tool catalog. */
type LifecycleCapability = "managed" | "unsupported" | "unknown";

interface Lifecycle {
  capability: LifecycleCapability;
}

interface Session {
  conversationId: string;
  interactionId: string;
  /** Kept so the session can be released later without the caller threading them back. */
  ctx: RequestContext;
  knId: string;
}

const START_TOOL = "bkn_start_interaction";
const FINISH_TOOL = "bkn_finish_interaction";

/**
 * Errors that mean "the session is gone, open a new one": a conversation swept
 * for being idle takes its interaction with it, and a long-lived client
 * outlives both.
 */
const STALE_SESSION_CODES = new Set([
  "conversation_required",
  // Trace Core's lifecycleError enum (agent-observability.yaml): a conversation
  // that was closed, swept, or never existed takes its interaction with it.
  "conversation_not_found",
  "conversation_closed",
  "conversation_expired",
  "interaction_required",
  "interaction_terminal",
  "interaction_in_progress",
]);

/**
 * Default attribution, so an SDK-opened conversation is identifiable in Trace.
 * A caller overrides it with `ClientOptions.agentName`; the contract asks for the
 * same name on every start within one conversation, so it is fixed per client.
 */
const DEFAULT_AGENT_NAME = "openbkn-sdk";

/** `startInteractionRequest.agent_name` is `maxLength: 128` (agent-observability.yaml). */
export const MAX_AGENT_NAME_LENGTH = 128;

function agentNameFor(ctx: RequestContext): string {
  const name = ctx.agentName?.trim() || DEFAULT_AGENT_NAME;
  // Checked locally: the server's `agent_name_invalid` would otherwise arrive on
  // every managed call, naming a handshake the caller never made.
  if ([...name].length > MAX_AGENT_NAME_LENGTH) {
    throw new InputError(
      `agentName must be at most ${MAX_AGENT_NAME_LENGTH} characters (got ${[...name].length}).`,
    );
  }
  return name;
}

/**
 * Releasing a session happens on the way out of a process, after the result is
 * already printed. The MCP transport has no timeout of its own, so without this
 * a wedged server would leave the command hanging with nothing left to say.
 */
const RELEASE_TIMEOUT_MS = 3_000;

const contracts = new Map<string, Promise<Lifecycle>>();

/**
 * How long a failed probe stays failed.
 *
 * Not caching it at all means a deploy whose `/mcp/info` is durably gone pays
 * a doomed round trip per business call — or worse, waits out the request
 * timeout each time, where before only the first call did. Caching it forever
 * lets one blip decide the rest of the process. A short window does both jobs.
 *
 * Recorded per caller, not per deploy. The probe carries `ctx.token`, so a 401
 * is a fact about one identity; filing it under the base URL alone would let
 * one expired token answer "no lifecycle here" for every other identity using
 * that host. Same distinction the session key draws — that one decides whose
 * session a request joins, this one decides whose failure it inherits.
 */
const PROBE_FAILURE_TTL_MS = 30_000;
const probeFailures = new Map<string, number>();
const sessions = new Map<string, Promise<Session>>();

/** Reset both caches. Tests only — a process never needs this. */
export function resetLifecycleCaches(): void {
  contracts.clear();
  probeFailures.clear();
  sessions.clear();
}

/**
 * Which contract does this deploy speak? Answered from the global tool catalog,
 * which is a plain GET — no MCP session, no KN, one call per process. An
 * unreachable or malformed catalog is distinct from a catalog that explicitly
 * lacks lifecycle tools: ordinary calls preserve the old best-effort fallback,
 * while receipt-required calls must not mistake uncertainty for legacy support.
 */
const MANAGED_LIFECYCLE: Lifecycle = { capability: "managed" };
const NO_LIFECYCLE: Lifecycle = { capability: "unsupported" };
const UNKNOWN_LIFECYCLE: Lifecycle = { capability: "unknown" };

/**
 * The raw catalog probe. Receipt callers need an authentication failure to stay
 * actionable, while ordinary calls preserve their longstanding best-effort
 * fallback through {@link lifecycleFor} below.
 */
function lifecycleProbeFor(ctx: RequestContext): Promise<Lifecycle> {
  const failureKey = `${ctx.baseUrl}\0${identityOf(ctx)}`;
  const failedAt = probeFailures.get(failureKey);
  if (failedAt !== undefined) {
    if (Date.now() - failedAt < PROBE_FAILURE_TTL_MS) return Promise.resolve(UNKNOWN_LIFECYCLE);
    probeFailures.delete(failureKey);
  }
  let pending = contracts.get(failureKey);
  if (!pending) {
    pending = mcpInfo(ctx).then((info): Lifecycle => {
      const names = toolNames(info);
      if (!names) {
        // A malformed 200 response is as inconclusive as a rejected probe. Do
        // not let it decide a long-lived process permanently, but avoid a new
        // doomed round trip for every business call in this short window.
        contracts.delete(failureKey);
        probeFailures.set(failureKey, Date.now());
        return UNKNOWN_LIFECYCLE;
      }
      return names.includes(START_TOOL) ? MANAGED_LIFECYCLE : NO_LIFECYCLE;
    });
    contracts.set(failureKey, pending);
    // Same rule the session cache follows: a probe that failed is not a lasting
    // answer about the deploy, only about that moment. Keeping it would let one
    // blip — a 502, a token mid-refresh — decide that every later call in this
    // process goes out without a `bkn_context`, and the reopen path cannot
    // recover from that, since it sees no context to reopen.
    pending.catch((err: unknown) => {
      contracts.delete(failureKey);
      // An authorization failure is not worth a window at all: a refresh can
      // fix it on the very next call, and holding it would spend 30s answering
      // from a credential that has already been replaced.
      if (!isAuthFailure(err)) probeFailures.set(failureKey, Date.now());
    });
  }
  return pending;
}

export function lifecycleFor(ctx: RequestContext): Promise<Lifecycle> {
  return lifecycleProbeFor(ctx).catch((): Lifecycle => UNKNOWN_LIFECYCLE);
}

function isAuthFailure(err: unknown): boolean {
  return err instanceof HttpError && (err.status === 401 || err.status === 403);
}

function toolNames(info: unknown): string[] | undefined {
  const tools = (info as { tools?: Array<{ name?: unknown }> } | undefined)?.tools;
  if (!Array.isArray(tools)) return undefined;
  return tools.flatMap((tool) => (typeof tool?.name === "string" ? [tool.name] : []));
}

export async function requireKnownLifecycleCapability(ctx: RequestContext): Promise<void> {
  try {
    if ((await lifecycleProbeFor(ctx)).capability !== "unknown") return;
  } catch (err) {
    // Do not turn a credential failure into a generic retry hint. HttpError
    // carries the CLI's established authentication diagnosis and exit status.
    if (isAuthFailure(err)) throw err;
  }
  throw new ToolError(
    "Context-loader lifecycle capability is unavailable; retry when the catalog can be read.",
    "lifecycle_capability_unknown",
  );
}

/**
 * A caller that already owns a conversation always wins: it has a real business
 * turn to bind evidence to, which is worth more than anything we could open.
 */
function callerOwnedSession(
  ctx: RequestContext,
): Pick<Session, "conversationId" | "interactionId"> | undefined {
  const conversationId = ctx.trace?.conversationId;
  const interactionId = ctx.trace?.interactionId;
  return conversationId && interactionId ? { conversationId, interactionId } : undefined;
}

function readId(result: unknown, field: string, tool: string): string {
  const value = (result as Record<string, unknown> | undefined)?.[field];
  if (typeof value !== "string" || !value) {
    throw new Error(`Managed lifecycle: ${tool} returned no ${field}.`);
  }
  return value;
}

/**
 * A conversation the caller named but did not pair with an interaction.
 *
 * `bkn_start_interaction` can open an interaction inside an existing
 * conversation, so this is a supported input, not an incomplete one — `--conversation-id` and
 * `--interaction-id` are independent flags. Opening a fresh conversation here
 * instead would file the evidence somewhere the caller never asked for, without
 * saying so.
 */
function callerNamedConversation(ctx: RequestContext): string | undefined {
  return ctx.trace?.interactionId ? undefined : ctx.trace?.conversationId;
}

/**
 * Could this failure be the deploy rejecting *this conversation*?
 *
 * The retry below is worth a second handshake only then. The rule is that a
 * failure which will repeat identically must not be paid for twice, and each
 * handshake opens its own MCP session — so with something stored, a deploy that
 * is down or a credential that has expired would otherwise cost every command
 * double.
 *
 * A refusal reaches here as a `ToolError` — an MCP `isError` result, or a
 * JSON-RPC error from a gateway that validates before dispatch — or as a 4xx.
 * Three 4xx are excluded: auth, because the credential is unchanged on the
 * retry, and 408/429, because they describe the request rather than its
 * arguments. Those two are also the only 4xx where a retry can *succeed*, which
 * would trade a conversation that was fine for a new one and overwrite whatever
 * the caller had stored.
 *
 * Everything else — 5xx, a transport failure, a malformed body, a missing
 * session id — is about reaching the deploy at all, never about which
 * conversation was named.
 */
function refusesThisConversation(err: unknown): boolean {
  if (err instanceof ToolError) return true;
  if (!(err instanceof HttpError) || isAuthFailure(err)) return false;
  if (err.status === 408 || err.status === 429) return false;
  return err.status >= 400 && err.status < 500;
}

/**
 * The conversation a fresh session should join, if any.
 *
 * A caller-named one and a remembered one look the same to the server; they
 * differ in what a failure means. `ensureSession` drops the remembered one and
 * tries again, because it offered convenience, not intent.
 */
function joinTarget(ctx: RequestContext): string | undefined {
  return callerNamedConversation(ctx) ?? ctx.rememberedConversationId;
}

async function openSession(ctx: RequestContext, knId: string, question: string): Promise<Session> {
  const named = joinTarget(ctx);
  // Without a conversation_id the server mints a fresh conversation, so only a
  // conversation being joined is passed in. `conversation_mode` is `continue`
  // exactly then and `new` otherwise; it, `question` and `agent_name` are
  // required on every start (context-loader mcp.yaml).
  const started = await callToolRaw(ctx, knId, START_TOOL, {
    question,
    conversation_mode: named ? "continue" : "new",
    agent_name: agentNameFor(ctx),
    ...(named ? { conversation_id: named } : {}),
  });
  return {
    ctx,
    knId,
    conversationId: readId(started, "conversation_id", START_TOOL),
    interactionId: readId(started, "interaction_id", START_TOOL),
  };
}

/**
 * A session belongs to one KN, one caller, and one named conversation.
 *
 * The KN because the conversation is opened over an MCP session bound to
 * `x-kn-id`; the identity because an embedder can drive one base URL with a
 * different token per user, and reusing a session across them would file one
 * user's evidence under another's business turn — and later release it with the
 * wrong credential. The token is hashed rather than stored: this key ends up in
 * a long-lived map, and a bearer token does not belong there.
 *
 * The named conversation because a caller that asked for `conv_B` must not be
 * handed an interaction on `conv_A`. The other two dimensions protect a caller
 * who said nothing; this one protects a caller who said something specific,
 * which is the worse of the two to ignore.
 *
 * Parts are joined on NUL, written as the two-character escape so the file
 * stays plain ASCII. A separator that cannot occur inside a part is the point:
 * with a space, `token="a b"` + `domain="c"` and `token="a"` + `domain="b c"`
 * hash the same. Unreachable in practice, but this key decides whose session a
 * request joins.
 */
function identityOf(ctx: RequestContext): string {
  return createHash("sha256").update(ctx.token).digest("hex").slice(0, 16);
}

function sessionKey(ctx: RequestContext, knId: string): string {
  // Both conversation inputs belong in the key for the same reason: a caller
  // that asked for one must not be handed an interaction on another.
  const wanted = ctx.trace?.conversationId ?? ctx.rememberedConversationId ?? "";
  return `${ctx.baseUrl}\0${knId}\0${identityOf(ctx)}\0${wanted}`;
}

/**
 * The session is cached as a promise so concurrent first calls share one
 * handshake — a conversation permits a single active interaction, so two racing
 * opens would leave one of them rejected.
 */
function ensureSession(ctx: RequestContext, knId: string, question: string): Promise<Session> {
  const key = sessionKey(ctx, knId);
  const cached = sessions.get(key);
  if (cached) return cached;
  // A remembered conversation may have been swept, or may still hold an active
  // interaction — a conversation permits one at a time. Either way it is this
  // caller's own convenience failing, so drop it and open a fresh one rather
  // than failing the command: without this, one unusable id would break every
  // later run, and the reopen path in `withManagedLifecycle` cannot help. That
  // path needs a context to retry with, and a handshake that throws prevents
  // `bknContextFor` from returning one to retry with.
  // "This session is joining the remembered conversation" — asked as the reason,
  // not as a value comparison. A caller that names the same id it also stored
  // still named it, and a named conversation must never be swapped out from
  // under them.
  const remembered = callerNamedConversation(ctx) ? undefined : joinTarget(ctx);
  const opening = remembered
    ? openSession(ctx, knId, question).catch((err: unknown) => {
        if (!refusesThisConversation(err)) throw err;
        const { rememberedConversationId: _dropped, ...fresh } = ctx;
        return openSession(inheritVersionCheck(ctx, fresh), knId, question);
      })
    : openSession(ctx, knId, question);
  // Report only a conversation this call minted. One the caller named is
  // already theirs to keep, and echoing it back would let a `--conversation-id`
  // meant for a single command quietly become the stored default.
  if (!callerNamedConversation(ctx)) {
    opening
      .then((session) => {
        // Joining a remembered conversation returns the same id; reporting it
        // would restamp its age and make "opened at" mean "last used at".
        if (session.conversationId === ctx.rememberedConversationId) return;
        ctx.onConversationOpened?.(session.conversationId);
      })
      .catch(() => {
        /* the handshake failure is surfaced by the caller awaiting `opening` */
      });
  }
  sessions.set(key, opening);
  // A failed handshake must not poison the cache for the rest of the process.
  opening.catch(() => sessions.delete(key));
  return opening;
}

function contextFor(session: Pick<Session, "conversationId" | "interactionId">): BknContext {
  return { conversation_id: session.conversationId, interaction_id: session.interactionId };
}

/**
 * A complete business context is authoritative for its operation. An
 * interaction-only trace header is a second, incomplete context channel, so
 * omit it for every path that supplies a full `bkn_context`.
 */
export function requestContextForBusinessContext(
  ctx: RequestContext,
  bknContext: BknContext | undefined,
): RequestContext {
  if (!bknContext || !ctx.trace?.interactionId || ctx.trace.conversationId) return ctx;
  const { interactionId: _interactionId, ...trace } = ctx.trace;
  return inheritVersionCheck(ctx, { ...ctx, trace });
}

/** Resolve a `bkn_context` for one call, or `undefined` when no managed contract exists. */
export async function bknContextFor(
  ctx: RequestContext,
  knId: string,
  question: string,
  requireKnownCapability = false,
): Promise<BknContext | undefined> {
  const lifecycle = await lifecycleFor(ctx);
  if (lifecycle.capability === "unknown") {
    if (requireKnownCapability) await requireKnownLifecycleCapability(ctx);
    return undefined;
  }
  if (lifecycle.capability !== "managed") return undefined;

  const owned = callerOwnedSession(ctx);
  if (owned) return contextFor(owned);

  return contextFor(await ensureSession(ctx, knId, question));
}

/**
 * The server's error code, wherever it landed: an HTTP `/kn/*` call carries it
 * in the response body, an MCP call in an `isError` result or a JSON-RPC
 * top-level error — both reach here as a `ToolError`.
 *
 * Including the JSON-RPC channel is deliberate and reaches past this PR's
 * subject: any managed tool call whose refusal names a stale-session code now
 * takes the same recovery as one reported through `isError`. Same meaning, same
 * path. It is bounded — one reopen, and a top-level error means the call was
 * never dispatched, so resending has nothing to undo — and a caller-owned
 * session is still never swapped out.
 */
function serverErrorCode(err: unknown): string | undefined {
  if (err instanceof ToolError) return err.code;
  if (!(err instanceof HttpError)) return undefined;
  try {
    // Context-loader REST errors are `ErrorCompact` and Core's lifecycle errors
    // are `lifecycleError`, both with a top-level `code`; older deploys nested
    // it under `error`.
    const parsed = JSON.parse(err.body) as { code?: unknown; error?: { code?: unknown } };
    if (typeof parsed.error?.code === "string") return parsed.error.code;
    return typeof parsed.code === "string" ? parsed.code : undefined;
  } catch {
    return undefined;
  }
}

/**
 * Run one `/kn/*` call under a managed session, reopening once if the session
 * died between calls. The retry is bounded to a single attempt: a second
 * failure is about the request, not the session.
 */
export async function withManagedLifecycle<T>(
  ctx: RequestContext,
  knId: string,
  question: string,
  send: (bknContext: BknContext | undefined, requestContext: RequestContext) => Promise<T>,
  requireKnownCapability = false,
): Promise<T> {
  const first = await bknContextFor(ctx, knId, question, requireKnownCapability);
  try {
    return await send(first, requestContextForBusinessContext(ctx, first));
  } catch (err) {
    const code = serverErrorCode(err);
    // Only a session we opened is ours to reopen. A caller-supplied
    // conversation belongs to its owner's turn; silently replacing it would
    // detach the evidence from the business conversation it was meant for.
    if (!first || callerOwnedSession(ctx) || !code || !STALE_SESSION_CODES.has(code)) throw err;
    sessions.delete(sessionKey(ctx, knId));
    const reopened = await bknContextFor(ctx, knId, question, requireKnownCapability);
    if (!reopened) throw err;
    return await send(reopened, requestContextForBusinessContext(ctx, reopened));
  }
}

/**
 * Release the interactions this process opened.
 *
 * `bkn_finish_interaction` takes the interaction id and an outcome. The
 * outcome is `cancelled` because a
 * CLI invocation has no answer artifact to close over, and `completed` without
 * one is rejected.
 *
 * Best-effort by construction: this runs while a process is shutting down,
 * where a throw would turn a successful command into a failed one. The whole
 * release is bounded, not just its final call — the handshake being awaited
 * here has no deadline of its own, so a server that accepted the connection
 * and went quiet would hold the process open after its output was printed.
 */
export async function releaseLifecycleSessions(): Promise<void> {
  const pending = [...sessions.values()];
  sessions.clear();
  await Promise.all(pending.map((opening) => withDeadline(releaseOne(opening))));
}

/** Resolve when `work` settles or the release deadline passes, whichever is first. */
function withDeadline(work: Promise<void>): Promise<void> {
  return new Promise((resolve) => {
    const timer = setTimeout(resolve, RELEASE_TIMEOUT_MS);
    // Do not hold the event loop open for a release nobody is waiting on.
    timer.unref?.();
    work.finally(() => {
      clearTimeout(timer);
      resolve();
    });
  });
}

async function releaseOne(opening: Promise<Session>): Promise<void> {
  try {
    const session = await opening;
    await callToolRaw(
      session.ctx,
      session.knId,
      FINISH_TOOL,
      {
        interaction_id: session.interactionId,
        outcome: "cancelled",
        reason: "client session ended",
      },
      undefined,
      RELEASE_TIMEOUT_MS,
    );
  } catch {
    // The interaction ages out on its own; a failed release is not the
    // command's problem.
  }
}

File: python/bkn_osdk/lifecycle.py
# Copyright (c) 2026 OpenBKN. All rights reserved.
# Licensed under the Apache License, Version 2.0. See the LICENSE file in the project root.

"""Managed lifecycle sessions — the thing that makes a read evidence.

Every MCP read wants a `bkn_context` naming the conversation and interaction it
belongs to, so the platform can file the operation against a turn instead of an
anonymous bucket. This module opens one interaction per `session(traced=True)`
scope, reuses it for every read inside, and finishes it on the way out.

One interaction per scope, not per query: opening one per read would cost a
round trip each time and shatter the evidence chain into unrelated turns, which
is the opposite of the point.

The deploy this was built against speaks the contract where
`bkn_start_interaction` alone mints both ids — there is no separate
`bkn_create_conversation` in its catalog — and `bkn_context` accepts only those
two ids.
"""

from __future__ import annotations

import threading
from collections.abc import Callable, Iterator
from contextlib import contextmanager
from contextvars import ContextVar
from dataclasses import dataclass, field
from typing import Any, TypeVar

from .config import Context
from .errors import BknError, HttpError, ToolError, required_action
from .mcp import call_tool, tool_catalog

__all__ = [
    "Interaction",
    "current_interaction",
    "ensure_interaction",
    "interaction_scope",
    "with_context_retry",
]

T = TypeVar("T")

START_TOOL = "bkn_start_interaction"
FINISH_TOOL = "bkn_finish_interaction"
#: The other contract in the wild: a conversation is created before an
#: interaction is started inside it.
CREATE_TOOL = "bkn_create_conversation"

DEFAULT_QUESTION = "bkn-osdk read"
#: Display-only attribution, so an SDK-opened turn is identifiable in Trace.
AGENT_NAME = "bkn-osdk"


@dataclass(frozen=True)
class Catalog:
    """What a deploy's tool catalog says about opening a turn.

    Two builds advertise the same tool names and still disagree on the
    arguments: one declares `conversation_mode` required on the start tool, the
    other never published it. Only the tool's own schema separates them, so the
    probe keeps it rather than throwing it away with the names.
    """

    tools: frozenset[str]
    #: True when the start tool declares `conversation_mode`.
    declares_conversation_mode: bool = False
    #: False when the catalog could not be read at all. An absent tool and an
    #: unreadable catalog look identical in `tools`, and they are not the same
    #: thing: one is a deploy that cannot open a turn, the other is a deploy
    #: this SDK knows nothing about yet.
    known: bool = True


@dataclass
class Interaction:
    """One open turn: the ids every traced read carries."""

    kn_id: str
    conversation_id: str
    interaction_id: str
    #: Receipts collected from the reads made inside this interaction.
    receipts: list[dict[str, Any]] = field(default_factory=list)
    #: True when the caller handed us this turn. Never finish someone else's.
    caller_owned: bool = False

    @property
    def bkn_context(self) -> dict[str, str]:
        """Exactly the two ids. This contract rejects anything else outright."""
        return {
            "conversation_id": self.conversation_id,
            "interaction_id": self.interaction_id,
        }


_current: ContextVar[dict[str, Interaction] | None] = ContextVar(
    "bkn_osdk_interactions", default=None
)
_lock = threading.Lock()
_catalogs: dict[str, Catalog] = {}


def current_interaction(ctx: Context, kn_id: str) -> Interaction:
    """The interaction for this scope and network, opening one if needed.

    Scoped per network: one `session(traced=True)` that reads two networks holds
    two interactions, because an interaction belongs to a knowledge network.
    """
    open_interactions = _current.get()
    if open_interactions is None:
        raise BknError(
            "A traced read needs an open interaction. Wrap the call in "
            "`with bkn_osdk.session(traced=True):`, which opens one and finishes it on exit."
        )
    existing = open_interactions.get(kn_id)
    if existing is not None:
        return existing

    interaction = _start(ctx, kn_id)
    open_interactions[kn_id] = interaction
    return interaction


def interaction_scope() -> ContextVar[dict[str, Interaction] | None]:
    """The scope's registry, so `session()` can open and close one."""
    return _current


def _start(ctx: Context, kn_id: str, question: str = DEFAULT_QUESTION) -> Interaction:
    """Join the caller's turn where there is one; otherwise open our own.

    A host that already owns a business turn — the sandbox injects both ids per
    execution — has something real to bind evidence to, which is worth more than
    anything opened here. Where only a conversation is named, the interaction is
    opened *inside* it rather than in a fresh conversation the caller never
    asked for.
    """
    if ctx.conversation_id and ctx.interaction_id:
        return Interaction(kn_id, ctx.conversation_id, ctx.interaction_id, caller_owned=True)

    catalog = _catalog(ctx)
    tools = catalog.tools
    if catalog.known and START_TOOL not in tools:
        raise BknError(
            f"This deploy's tool catalog has no {START_TOOL}, so no managed interaction can "
            "be opened. Read without `traced=True`; the REST path needs no session."
        )
    if catalog.known and CREATE_TOOL in tools:
        # The older contract mints the conversation separately. Nothing here has
        # been able to test it, so it is refused rather than guessed at.
        raise BknError(
            f"This deploy speaks the {CREATE_TOOL} lifecycle contract, which this runtime "
            "does not implement yet. Read without `traced=True`."
        )

    result = _opened(
        ctx,
        kn_id,
        {
            "question": question,
            #: Display-only, but the only thing separating a turn this SDK
            #: opened from a real agent's in a Trace listing.
            "agent_name": AGENT_NAME,
            **({"conversation_id": ctx.conversation_id} if ctx.conversation_id else {}),
            # Sent only where the tool declares it: this contract validates
            # strictly, and a deploy may reject an argument it never published.
            # An unreadable catalog answers "no" — send what has always worked
            # rather than guess a new field in.
            **(
                {"conversation_mode": "continue" if ctx.conversation_id else "new"}
                if catalog.declares_conversation_mode
                else {}
            ),
        },
    )
    value = result.value if isinstance(result.value, dict) else {}
    conversation_id = value.get("conversation_id")
    interaction_id = value.get("interaction_id")
    if not isinstance(conversation_id, str) or not isinstance(interaction_id, str):
        raise BknError(f"{START_TOOL} returned no conversation/interaction id: {value}")
    return Interaction(kn_id, conversation_id, interaction_id)


def _opened(ctx: Context, kn_id: str, arguments: dict[str, Any]) -> Any:
    """Open the turn, and say what a failure to open one actually costs.

    The tool's own message is true and unhelpful on its own — a caller who asked
    for a search reads `trace_core_unavailable` and cannot tell whether the SDK,
    the network, or the platform is at fault, nor what still works. Most of the
    read surface does not need a turn at all, and that is the useful half of the
    answer.
    """
    from .mcp import call_tool

    try:
        return call_tool(ctx, kn_id, START_TOOL, arguments)
    except ToolError as error:
        raise ToolError(
            error.code,
            f"This deploy could not open a managed turn — {error.message} "
            "Calls that require one — search, the capability routes, `traced=True` — "
            "cannot run until it can. Typed reads of object types, subgraphs and metrics "
            "need no turn and are unaffected.",
            required_action=error.required_action,
            retryable=error.retryable,
            retry_after_ms=error.retry_after_ms,
        ) from error


@contextmanager
def ensure_interaction(ctx: Context, kn_id: str) -> Iterator[Interaction]:
    """A turn to attach a call to: the one in scope, or a new one.

    Three cases, and the difference that matters is who closes it:

    * inside a `session(traced=True)` scope — the scope's own turn, left open
      for the scope to finish, so the evidence stays together;
    * where the caller named one, as the sandbox does through the environment —
      joined and never finished, because ending someone else's business turn
      early is not this SDK's call to make;
    * otherwise — opened here and finished on exit, which costs a round trip.

    `current_interaction` is the same thing without the last case: it raises
    where there is no scope, rather than opening one.
    """
    scope = _current.get()
    if scope is not None:
        yield current_interaction(ctx, kn_id)
        return

    interaction = _start(ctx, kn_id)
    try:
        yield interaction
    except BaseException:
        # A turn closed as `completed` after the call it exists for raised is a
        # false entry in the evidence chain — and every capability route now
        # opens one of these, so the falsehood would be the common case rather
        # than the rare one. `session(traced=True)` already distinguishes them.
        finish(ctx, interaction, "failed", None)
        raise
    else:
        finish(ctx, interaction, "completed", None)


def with_context_retry(ctx: Context, kn_id: str, send: Callable[[dict[str, str] | None], T]) -> T:
    """Attach the turn where there is one; otherwise send bare and adapt.

    The platform's own rule for its REST capability layer is one line: name a
    session and the call is managed, name none and it is ad hoc. This follows it.

    **Where a turn exists, every call carries it.** Inside a `session(traced=True)`
    scope, or when the caller handed one in — the sandbox injects both ids per
    execution — the context goes out on the first attempt. Waiting for a refusal
    would silently drop the evidence on any deploy that does not demand one, and
    a traced scope asked for exactly that evidence.

    **Where none exists, none is minted.** A read outside a traced scope is a
    capability call, not an agent turn; opening a session to satisfy a guard
    would file a single-operation record that dilutes the concept rather than
    documenting anything. So the call goes bare — and only where a deploy still
    refuses it is a short-lived turn opened for the retry.
    """
    if _has_turn(ctx):
        with ensure_interaction(ctx, kn_id) as interaction:
            return send(interaction.bkn_context)

    try:
        return send(None)
    except (HttpError, ToolError) as error:
        if not _needs_context(error):
            raise

    with ensure_interaction(ctx, kn_id) as interaction:
        return send(interaction.bkn_context)


def _has_turn(ctx: Context) -> bool:
    """Whether there is a turn to attach to without minting one."""
    return ctx.traced or bool(ctx.conversation_id and ctx.interaction_id)


def _needs_context(error: HttpError | ToolError) -> bool:
    """Whether this refusal is the lifecycle middleware asking for a session."""
    if isinstance(error, ToolError):
        return error.code in _LIFECYCLE_ACTIONS or error.required_action in _LIFECYCLE_ACTIONS
    return required_action(error.body) in _LIFECYCLE_ACTIONS


#: What a deploy answers when it wants a `bkn_context`. The two contracts name
#: different actions for the same requirement, so both are recognised.
_LIFECYCLE_ACTIONS = frozenset(
    {"conversation_required", "create_conversation", "start_interaction", START_TOOL}
)


def finish(ctx: Context, interaction: Interaction, outcome: str, answer: str | None) -> None:
    """Close the turn, unless the caller owns it.

    Finishing someone else's interaction would end their business turn early —
    the host that opened it decides when it is done.

    Our own failures are swallowed: the reads already happened and their receipts
    are already recorded, so turning a bookkeeping error into the caller's
    exception would hide the result they came for.
    """
    if interaction.caller_owned:
        return
    try:
        call_tool(
            ctx,
            interaction.kn_id,
            FINISH_TOOL,
            {
                "interaction_id": interaction.interaction_id,
                "outcome": outcome,
                **({"answer": answer} if answer and outcome == "completed" else {}),
            },
        )
    except BknError:
        return


def _catalog(ctx: Context) -> Catalog:
    """The deploy's tool catalog, read once per process per deploy."""
    with _lock:
        cached = _catalogs.get(ctx.base_url)
    if cached is not None:
        return cached

    try:
        payload = tool_catalog(ctx)
    except BknError:
        # `/mcp/info` is a convenience, not the read path. A deploy that does not
        # serve it is one this SDK cannot describe — so it sends what has always
        # worked and lets the tool itself refuse if it must, rather than turning
        # every traced read into an error about a probe. Not cached: a broken
        # probe is usually temporary, and the next call should find out.
        return Catalog(tools=frozenset(), known=False)

    entries = payload.get("tools") if isinstance(payload, dict) else None
    tools = [entry for entry in (entries or []) if isinstance(entry, dict)]
    names = frozenset(entry["name"] for entry in tools if isinstance(entry.get("name"), str))
    catalog = Catalog(
        tools=names,
        declares_conversation_mode=any(
            entry.get("name") == START_TOOL and _declares(entry, "conversation_mode")
            for entry in tools
        ),
    )
    with _lock:
        _catalogs[ctx.base_url] = catalog
    return catalog


def _declares(tool: dict[str, Any], field: str) -> bool:
    """Whether a tool's published input schema names this argument.

    Keyed on `properties` rather than `required`: a field the deploy published
    is one it knows, and sending a published-but-optional argument is safe where
    sending an unpublished one is not. The platform is moving the other way
    anyway — a build that listed `conversation_mode` as optional now lists it as
    required.
    """
    schema = tool.get("input_schema") or tool.get("inputSchema")
    if not isinstance(schema, dict):
        return False
    properties = schema.get("properties")
    return isinstance(properties, dict) and field in properties


def _reset_for_tests() -> None:
    with _lock:
        _catalogs.clear()
